libflux/src/ast/walk/mod.rs

```diff
@@ -231,109 +231,114 @@ impl<'a> Node<'a> {
 // Done is called on that Node to signal that we are done with that Node.
 //
 // If Visit returns None, walk will not recurse on the children.
-pub trait Visitor: Sized {
+//
+// Note: the Rc in visit is to allow for multiple ownership of a node, i.e.
+//       a visitor can own a node as well as the walk funciton. This allows
+//       for nodes to persist outside the scope of the walk function and to
+//       be cleaned up once all owners have let go of the reference.
+pub trait Visitor<'b>: Sized {
     // Visit is called for a node.
     // The returned visitor will be used to walk children of the node.
     // If visit returns None, walk will not recurse on the children.
-    fn visit(&self, node: &Node) -> Option<Self>;
+    fn visit<'a>(&self, node: Rc<Node<'b>>) -> Option<Self>;
```

Comment:
nit, can you remove the `<'a>` from the function since its not used and then rename `'b` to `'a` for consistency?